## Loops & Loop operation's

In Python, a loop is a control structure that allows you to repeatedly execute a block of code multiple times. Loops are essential for automating repetitive tasks and iterating over collections of data, such as lists or strings.

Python provides two main types of loops i.e a "for" loop and a "while" loop.

For Loop: A 'for' loop is used to iterate over a sequence (such as a list, tuple, string, or range) and execute a block of code for each item in the sequence. It is typically used when you know the number of iterations in advance or when you want to iterate over elements one by one.

While Loop: A while loop is used to repeatedly execute a block of code as long as a specified condition is true. It is often used when you don't know in advance how many iterations are needed or when you want to continue looping until a specific condition is met.

In [1]:
numbers=[19,24,36,42,3,1,7,9,4,5,4,101]
#from this list of numbers print even and odd.

In [13]:
for i in numbers:
    if i%2==0:
        print("even ",i)
    else:
        print("odd ",i)

odd  19
even  24
even  36
even  42
odd  3
odd  1
odd  7
odd  9
even  4
odd  5
even  4
odd  101


In [6]:
type(len(numbers))

int

In [7]:
type(numbers)

list

In [2]:
for i in range(len(numbers)):
    numbers[i]
    if i%2==0:
        print("even ", numbers[i])
    else:
        print("odd ", numbers[i])

even  19
odd  24
even  36
odd  42
even  3
odd  1
even  7
odd  9
even  4
odd  5
even  4
odd  101


#### Prints all the numbers between 100 and 1000 (inclusive) that are divisible by both 5 and 6, and  prints them in rows of 10 numbers per row.

In [4]:
count=0 
for i in range(100,1001):#we use a for loop to iterate through numbers from 100 to 1000 (inclusive) using the range() function.
    if i%5==0 and i%6==0:
        print(i,end=" ")
        count+=1 #Increment the count variable by 1 to keep track of the number of numbers printed.
        if count%10==0: # If count is a multiple of 10, it means that 10 numbers have been printed on the current line.
                print() #so we print a newline character print() to start a new line.

120 150 180 210 240 270 300 330 360 390 
420 450 480 510 540 570 600 630 660 690 
720 750 780 810 840 870 900 930 960 990 


In [8]:
#This is one more way to solve the above problem.
count = 1
for i in range(100,1001):
    if i % 5==0 and i % 6 ==0:
        if count%10 != 0:
            print(i, end = " ")
        else:
            print(i)
        count = count+1

120 150 180 210 240 270 300 330 360 390
420 450 480 510 540 570 600 630 660 690
720 750 780 810 840 870 900 930 960 990


#### Write a program that prompts the user to enter the number of students and each student’s score, and displays the highest score

In [11]:
# Method 1
students=int(input("Enter number of students  "))
name=input("Enter name of student  ")
score=int(input("Enter score of student  "))
for i in range(students-1):
    name_1=input("Enter name of student  ")
    score_1=int(input("Enter score of student  "))
    if score_1 > score:
        score=score_1
        name=name_1
print(name ,"is the highest scorer with score ",score)

Enter number of students  3
Enter name of student  jay
Enter score of student  10
Enter name of student  shree
Enter score of student  20
Enter name of student  ram
Enter score of student  16
shree is the highest scorer with score  20


In [12]:
#Method 2
students=int(input("Enter number of students  "))
score=0
name=""
for i in range(students):
    name_1=input("Enter name of student  ")
    score_1=int(input("Enter score of student  "))
    if score_1 > score:
        score=score_1
        name=name_1
print(name ,"is the highest scorer with score ",score)

Enter number of students  3
Enter name of student  a
Enter score of student  19
Enter name of student  b
Enter score of student  12
Enter name of student  c
Enter score of student  17
a is the highest scorer with score  19


#### Using a while loop find even and odd number form a  given list.

In [12]:
numbers=[19,24,3,12,56,4,23,10,2,17,101]

In [13]:
 len(numbers) < len(numbers)+1

True

In [15]:
i=0
while i < len(numbers):
    if numbers[i] % 2==0:
        print(numbers[i] ," is even number")
    else:
        print(numbers[i] ," is odd number")
    i+=1
    


19  is odd number
24  is even number
3  is odd number
12  is even number
56  is even number
4  is even number
23  is odd number
10  is even number
2  is even number
17  is odd number
101  is odd number


#### Write a Python Program to Find the Sum of Natural Numbers?

In [7]:
numbers=1
sum=0
limit=50
while numbers<= limit:
    sum+=numbers
    numbers+=1
print(sum)

1275


#### Write a program that reads an integer and displays all its smallest factors, also known as prime factors.
#### For example, if the input integer is 120, the output should be as follows: 2 2 2 3 5

In [5]:
a = int(input())
b = 2
l=[]
while b<=a:
    if a%b==0:
        c = a/b
        a = c
        l.append(b)
    else:
        b = b+1
print(l)


120
[2, 2, 2, 3, 5]


#### Write a Program to check whether the number is Armstrong or not?

An Armstrong number (also known as a narcissistic number or a pluperfect digital invariant) is a specific type of number in mathematics. It's a number that is equal to the sum of its own digits, each raised to the power of the number of digits in that number. 

For example:

153 is an Armstrong number because 1^3 + 5^3 + 3^3 = 153.

9474 is an Armstrong number because 9^4 + 4^4 + 7^4 + 4^4 = 9474.

In [10]:
number=int(input("Enter the number for which you want the number is armstrong number  "))
power = int(input("Enter the power (e.g., 3 for cubes): "))
temp=number
sum_1=0
while temp>0:
    digit=temp%10
    sum_1=sum_1+digit**power
    temp=temp//10
if sum_1==number:
    print(number," is a armstrong number ")
else:
    print(number," is not a armstrong number ")

Enter the number for which you want the number is armstrong number  9474
Enter the power (e.g., 3 for cubes): 4
9474  is a armstrong number 


#### Python Program to Find LCM between two numbers.

In [2]:
a=int(input("Enter the first number "))
b=int(input("Enter the second number "))
max_num= max(a,b)
while (True):
    if max_num%a==0 and max_num%b==0:
        break
    max_num += 1  # Update max_num by 1 in each iteration

print("The LCM of", a ,"and", b ,"is", max_num)

Enter the first number 12
Enter the second number 36
The LCM of 12 and 36 is 36


### Break Statements

In Python, the break statement is used to exit or terminate a loop prematurely, before the loop's normal termination condition is met. It is often used in situations where you want to exit a loop early based on some condition.

In [4]:
count=0
while count < 10:
    print("count:", count)
    count += 1 
    if count == 9: #When count becomes 9,condition becomes true.Once the condition is true, the break statement is executed.
        break      #This immediately exits the while loop, even though the loop condition (count < 10) would still be true.
                   #As a result, the loop is terminated, and "count: 9" is not printed.

count: 0
count: 1
count: 2
count: 3
count: 4
count: 5
count: 6
count: 7
count: 8


### Continue Statements

In Python, the continue statement is used to skip the rest of the current iteration of a loop (typically a for or while loop) and move to the next iteration. It allows you to bypass specific code within the loop for certain conditions without exiting the entire loop.

In [4]:
count=0
while count < 10:
    count += 1 
    if count == 7: 
        continue 
    print("count:", count)
#When count becomes 7,condition becomes true.
#When the condition is true, the continue statement is executed.
#This causes the loop to immediately move to the next iteration, skipping the remaining code in the current iteration.
#As a result, "count: 7" is not printed, and the loop continues with count set to 8.
#The loop continues until count reaches 10, at which point the loop condition (count < 10) becomes false, and the loop terminates.

count: 1
count: 2
count: 3
count: 4
count: 5
count: 6
count: 8
count: 9
count: 10


# Functions

Function:-In Python, a function is a reusable block of code that performs a specific task or set of tasks. Functions are used to organize code into logical and modular units, making it easier to understand, reuse, and maintain your code.

Functions are defined using the 'def' keyword followed by a name and parentheses containing any input parameters.

Parameters are optional, and they represent values that can be passed into the function.

Functions can optionally return a result using the 'return' statement. If no 'return' statement is used, the function returns 'None' by default.

Functions can be reused multiple times in your code, promoting code reusability and readability

Functions help break down a program into smaller, manageable parts, making code organization and maintenance easier,


In [ ]:
#Syntax to write a Function:-

def function_name(parameters):
    # Function body or code block
    # ...
    return result  # Optional return statement


In [72]:
def shriya_s(list_of_num):
    for i in list_of_num:   
        if i%2==0 and i%5==0:
            print(i, "gets blue t-shirt")
        else:
            print(i, "gets no t-shirt")

In [73]:
shriya_s([1,2,12,14,13,10,20,25,30,35,40])

1 gets no t-shirt
2 gets no t-shirt
12 gets no t-shirt
14 gets no t-shirt
13 gets no t-shirt
10 gets blue t-shirt
20 gets blue t-shirt
25 gets no t-shirt
30 gets blue t-shirt
35 gets no t-shirt
40 gets blue t-shirt


#### A lambda function is a small, anonymous (unnamed) function defined using the lambda keyword. Lambda functions are also sometimes referred to as "anonymous functions" or "lambda expressions." Unlike regular functions defined with the def keyword, lambda functions are typically used for short, simple operations.

Syntax:-
     
     lambda arguments: expression


In [44]:
add = lambda x,y: x + y
result = add(3, 5)  # Calls the lambda function with arguments 3 and 5
print(result)  # Prints 8

8


In [38]:
even = lambda x:x%2==0

In [39]:
even(8)

True

#### Given a list of ints, return True if the array contains a 3 next to a 3 somewhere.


In [41]:
#Functions is initialized by using def keyword.
def shriya_1(list_1):
    for i in range(len(list_1)-1):
        if list_1[i]==3 and list_1[i+1]==3:
            return True
        

In [42]:
shriya_1([1,2,3,3,4,6,56,3,3])

True

In [11]:
type(shriya_1)

function

In [45]:
def shriya_3(ls):
    for i in range(len(ls) - 1):
        if ls[i] == 3 and ls[i+1] == 3:
            return True
    return False

In [56]:
list1 = [1,2,34,5,6,7]
for i in list1:
    print(list1[i], list1[i+1])
#Hence the first two iteration are possible because the value is 1 and 2 which will give the index value at 1 and 2 since it is present but in 3rd iteration the value is 34 and there's no element at index 34 in the list, so this will raise an "IndexError" because it's trying to access an element that doesn't exist.

2 34
34 5


IndexError: list index out of range

In [71]:
#To avoid the "IndexError" and to correctly print pairs of consecutive elements from the list, we should use this  approach.
#This code will iterate through the indices of the list and print pairs of elements correctly.

list2=[2,3,4,5,67]
len(list2)
for i in range(len(list2)-1):
        print (list2[i],list2[i+1])

2 3
3 4
4 5
5 67


In [49]:
#The below code will prints each element of the string.
for i in 'shriya':
    print(i)

s
h
r
i
y
a


In [38]:
#The below code prints each element of the string along with its corresponding index.
a='shriya'
for i in range(len(a)):
    print(a[i],i)

s 0
h 1
r 2
i 3
y 4
a 5


In [80]:
#The below code prints each element of the list along with its corresponding index.
b=[1,2,3,4,5,6]
for i in range(len(b)):
    print(i,b[i])

0 1
1 2
2 3
3 4
4 5
5 6


#### Write a program that will prints all possible permutations of the elements from the given list without repetitions

In [1]:
my_list=[1,2,3]
for i in range(len(my_list)):
    for j in range(len(my_list)):
        for k in range(len(my_list)):
            if i!=j and j!=k and k!=i:
                print(my_list[i],my_list[j],my_list[k])
    

1 2 3
1 3 2
2 1 3
2 3 1
3 1 2
3 2 1
